core/src/stored/backends/dplcompat_device.cc

```diff
@@ -0,0 +1,418 @@
+            .and_then(get_converter("iothreads", io_threads_))
+            .and_then(get_converter("ioslots", io_slots_))
+            .and_then(get_converter("retries", retries_))
+            .and_then(get_converter("chunksize", chunk_size_))
+            .and_then(get_converter("program", program))
+            .and_then(get_converter("program_timeout", program_timeout));
+      !conversion_result) {
+    return tl::unexpected(conversion_result.error());
+  }
+
+  if (program.empty()) {
+    return tl::unexpected("Option 'program' is required\n"s);
+  }
+
+  if (auto result = m_storage.set_program(program); !result) { return result; }
+
+  if (program_timeout > 0) { m_storage.set_program_timeout(program_timeout); }
+
+  if (auto supported_options = m_storage.get_supported_options()) {
+    for (const auto& option_name : *supported_options) {
+      if (auto value = fetch_value(options, option_name);
+          value && !m_storage.set_option(option_name, *value)) {
+        return tl::unexpected(fmt::format("Error setting option '{}' to '{}'\n",
+                                          option_name, *value));
+      }
+    }
+  } else {
+    return tl::unexpected(
+        fmt::format("Cannot get supported options.\nCause: {}\n",
+                    supported_options.error()));
+  }
+
+  // OptionConsumer should have consumed all options at this point
+  if (!options.empty()) {
+    BStringList option_names;
+    for (const auto& [name, value] : options) { option_names.push_back(name); }
+    return tl::unexpected(fmt::format("Unknown options encountered: {}\n",
+                                      option_names.Join(", ")));
+  }
+  return {};
+}
+
+bool DropletCompatibleDevice::CheckRemoteConnection()
+{
+  Dmsg0(debug_trace, "CheckRemoteConnection called\n");
+  return setup() && m_storage.test_connection();
+}
+
+bool DropletCompatibleDevice::FlushRemoteChunk(chunk_io_request* request)
+{
+  const std::string obj_name = request->volname;
+  const std::string obj_chunk = get_chunk_name(request);
+  if (request->wbuflen == 0) {
+    Dmsg1(debug_info, "Not flushing empty chunk %s/%s\n", obj_name.c_str(),
+          obj_chunk.c_str());
+    return true;
+  }
+  Dmsg1(debug_trace, "Flushing chunk %s/%s\n", obj_name.c_str(),
+        obj_chunk.c_str());
+
+  auto inflight_lease = getInflightLease(request);
+  if (!inflight_lease) {
+    Dmsg0(debug_info, "Could not acquire inflight lease for %s %s\n",
+          obj_name.c_str(), obj_chunk.c_str());
+    return false;
+  }
+
+
+  /* Check on the remote backing store if the chunk already exists.
+   * We only upload this chunk if it is bigger then the chunk that exists
+   * on the remote backing store. When using io-threads it could happen
+   * that there are multiple flush requests for the same chunk when a
+   * chunk is reused in a next backup job. We only want the chunk with
+   * the biggest amount of valid data to persist as we only append to
+   * chunks. */
```

Comment:
What happens if i do a truncate & immediately after the volume is used for a backup ? Do we make sure to flush everything after a truncate ?

Reply:
Truncate is synchronous and will only return when all chunks were deleted from the backend storage.